src/IR.cpp

```diff
@@ -542,6 +542,144 @@ Expr Variable::make(Type type, std::string name, Buffer<> image, Parameter param
     return node;
 }
 
+Expr Shuffle::make(const std::vector<Expr> &vectors,
+                   const std::vector<int> &indices) {
+    internal_assert(!vectors.empty()) << "Shuffle of zero vectors.\n";
+    internal_assert(!indices.empty()) << "Shufle with zero indices.\n";
+    Type element_ty = vectors.front().type().element_of();
+    int input_lanes = 0;
+    for (Expr i : vectors) {
+        internal_assert(i.type().element_of() == element_ty) << "Shuffle of vectors of mismatched types.\n";
+        input_lanes += i.type().lanes();
+    }
+    for (int i : indices) {
+        internal_assert(0 <= i && i < input_lanes) << "Shuffle vector index out of range: " << i << "\n";
+    }
+
+    Shuffle *node = new Shuffle;
+    node->type = element_ty.with_lanes(static_cast<int>(indices.size()));
+    node->vectors = vectors;
+    node->indices = indices;
+    return node;
+}
+
+Expr Shuffle::make_interleave(const std::vector<Expr> &vectors) {
+    internal_assert(!vectors.empty()) << "Interleave of zero vectors.\n";
+
+    if (vectors.size() == 1) {
+        return vectors.front();
+    }
+
+    int lanes = vectors.front().type().lanes();
+
+    for (Expr i : vectors) {
+        internal_assert(i.type().lanes() == lanes)
+            << "Interleave of vectors with different sizes.\n";
+    }
+
+    std::vector<int> indices;
+    for (int i = 0; i < lanes; i++) {
+        for (int j = 0; j < static_cast<int>(vectors.size()); j++) {
+            indices.push_back(j * lanes + i);
+        }
+    }
+
+    return make(vectors, indices);
+}
+
+Expr Shuffle::make_concat(const std::vector<Expr> &vectors) {
+    internal_assert(!vectors.empty()) << "Concat of zero vectors.\n";
+
+    if (vectors.size() == 1) {
+        return vectors.front();
+    }
+
+    std::vector<int> indices;
+    int lane = 0;
+    for (int i = 0; i < static_cast<int>(vectors.size()); i++) {
+        for (int j = 0; j < vectors[i].type().lanes(); j++) {
+            indices.push_back(lane++);
+        }
+    }
+
+    return make(vectors, indices);
+}
+
+Expr Shuffle::make_slice(Expr vector, int begin, int stride, int size) {
+    if (begin == 0 && size == vector.type().lanes() && stride == 1) {
+        return vector;
+    }
+
+    std::vector<int> indices;
+    for (int i = 0; i < size; i++) {
+        indices.push_back(begin + i * stride);
+    }
+
+    return make({vector}, indices);
+}
+
+bool Shuffle::is_interleave() const {
```

Comment:
Timing test_apps indicated no change in performance, I think that the case where a shuffle instruction is what we are looking for, the cost of checking is minor, and in the case where it is not, it will probably fail quickly.